[6주차|과제1] 레스토랑 상세 페이지 구현하기 코드 리뷰 부탁드립니다.

jiunkoo opened the pull request:
선 PR 후 진행 하겠습니다.

1. Home, About, Restaurants, NotFound 페이지 만들기(완료)
2. window.location을 활용하여 App에서 라우팅 구현(완료)
3. react-router-dom을 활용하여 라우팅 구현(완료)
4. 테스트 커버리지 100% 달성(완료)

jiunkoo committed:
initial commit

jiunkoo committed:
App에서 RestaurantsPage 분리

jiunkoo committed:
[추가] AboutPage 테스트 추가

jiunkoo committed:
[구현] AboutPage 테스트 구현

jiunkoo committed:
[추가] HomePage 테스트 추가

jiunkoo committed:
[구현] HomePage 테스트 구현

jiunkoo committed:
[추가] NotFound 페이지 테스트 추가

jiunkoo committed:
[구현] NotFound 페이지 테스트 구현

jiunkoo committed:
[리팩토링] App에 HomePage, AboutPage, NotFoundPage 추가 및 간단한 라우팅 적용

jiunkoo committed:
[리팩토링] Homepage의 href를 Link로 변경

hannut91 commented:
잘 진행하고 계신 것 같아요 ㅎㅎ 계속 진행해주세요 😀

jiunkoo committed:
[리팩토링] 라우터 사용을 위해 BrowerRouter 컴포넌트로 Homepage 테스트 리팩토링

jiunkoo committed:
[리팩토링] App의 winidow.location 라우팅을 react-router-dom 활용 라우팅으로 변경

jiunkoo committed:
[리팩토링] react-router-dom 사용을 위해 MemoryRouter 컴포넌트로 App 테스트 리팩토링

jiunkoo committed:
[리팩토링] 라우터 사용을 위해 App을 BrowserRouter 컴포넌트로 리팩토링

jiunkoo committed:
[추가] App에서 구현한 Homepage 라우터 테스트 추가

jiunkoo committed:
[추가] App에서 구현한 About, Restaurants, NotFound 라우터 추가

jiunkoo committed:
[리팩토링] App에 헤더 추가 및 각 페이지 글자 크기 조정

jiunkoo committed:
historyApiFallback 설정을 webpack에 추가

jiunkoo committed:
[리팩토링] RestaurantsContainer 식당 목록에 Link 추가 및 테스트 리팩토링

jiunkoo committed:
[추가] api에 fetchRestaurant 테스트 추가

jiunkoo committed:
[구현] api에 fetchRestaurant 테스트 및 mock 구현

jiunkoo committed:
fixtures에 restaurant 테스트 파일 추가

jiunkoo committed:
[추가] RestaurantDetail에 레스토랑 상세 컴포넌트 테스트 추가

jiunkoo committed:
[구현] RestaurantDetail에 레스토랑 상세 컴포넌트 테스트 구현

hannut91 commented on src/App.jsx:
```diff
@@ -1,32 +1,28 @@
-      <CategoriesContainer />
-      <RestaurantsContainer />
+      <header>
+        <Link to="/"><h1>헤더</h1></Link>
```
크게 상관 없긴 한데 h1태그안에 Link태그가 들어간게 더 자연스러운 것 같아요. 헤더 텍스트안에서 Link할 수 있는 텍스트가 존재하는거니까요

hannut91 commented on src/App.test.jsx:
```diff
@@ -3,32 +3,70 @@ import React from 'react';
-  }));
+  context('with path /restaurants', () => {
+    it('renders the restaurants page', () => {
+      const { container } = render((
+        <MemoryRouter initialEntries={['/restaurants']}>
+          <App />
+        </MemoryRouter>
+      ));
```
반복적으로 사용되는 부분은 함수로 추출하면 더 간단하게 표현할 수 있을거에요

hannut91 commented on src/RestaurantDetail.test.jsx:
```diff
@@ -0,0 +1,37 @@
+import RestaruantDetail from './RestaurantDetail';
+
+test('RestaurantDetail', () => {
+  const restaurant = {
```
`fixture`를 만들었으니 활용하면 좋을 것 같네요

jiunkoo replied:
저번 과제 피드백으로 테스트에 들어가는 인자를 볼 수 있으면 더 좋을 것 같다는 피드백을 받았었는데 fixture를 활용하는게 더 좋을까요?

hannut91 replied:
흠 그렇다면 그냥 놔둬도 될 것 같아요. 의도만 있으면 되니까요. 제가 생각하기엔 어떤 데이터가 변경되더라도 fixture하나에만 의존하면 다른 테스트들이 전부 fixture에만 의존하니까 관리하기 편하다고 생각했었어요.

hannut91 commented on webpack.config.js:
```diff
@@ -14,4 +14,10 @@ module.exports = {
   resolve: {
     extensions: ['.js', '.jsx'],
   },
+  output: {
+    publicPath: '/',
+  },
```
`publicPath`를 이용해서 문제를 해결하셨네요. index.html의 `<head>`에 `<base href="/" />`를 입력하여 해결할 수도 있습니다  

See also
  - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

jiunkoo replied:
레스토랑 상세 페이지를 볼 수 없어서 설정을 다시 변경하였습니다 흑흑

hannut91 reviewed (changes requested):
잘 진행하고 계십니다. e2e테스트 결과를 보고 어떤 것을 더 구현해야하는지 확인해보면서 진행해주세요

jiunkoo committed:
[추가] actions에 Restaurant 상세 정보 불러오는 테스트 추가

jiunkoo committed:
[구현] actions에 Restaurant 상세 정보 불러오는 테스트 구현

jiunkoo committed:
[추가] reducer에 restaurant 상태 변경 테스트 추가

jiunkoo committed:
[추가] reducer에 restaurant 상태 변경 테스트 구현

jiunkoo committed:
[추가] RestaurantContainer에 레스토랑 상세 정보를 dispatch하는 테스트 추가

jiunkoo committed:
[구현] RestaurantContainer에 레스토랑 상세 정보를 dispatch하는 테스트 구현

jiunkoo committed:
[추가] 레스토랑 상세 페이지를 보여줄 restaurantDetailPage 테스트 추가

jiunkoo committed:
[구현] 레스토랑 상세 페이지를 보여줄 restaurantDetailPage 테스트 구현

jiunkoo committed:
[리팩토링] App에 restaurantDetailPage 경로 추가

jiunkoo committed:
[리팩토링] restaurant 데이터 불러올 때까지 로딩 문구 띄우도록 테스트 리팩토링

jiunkoo committed:
레스토랑 상세 페이지 바로 나오도록 webpack 설정 변경

jiunkoo commented:
코드 구현은 끝냈고, 리팩토링은 퇴근 후 진행하겠습니다.

해야할 것
- RestaurantDetailContainer와 RestaurantDetail 페이지&테스트 합치기
- 코드 리뷰 반영

hannut91 commented on src/RestaurantDetailPage.test.jsx:
```diff
@@ -0,0 +1,77 @@
+      expect(queryByText(/팔보채/)).not.toBeNull();
+    });
+
+    it('숫자가 아닌 경우 에러 페이지를 띄운다', () => {
```
이렇게 하는건 어떨까요?
```js
context('param에 restaurantId가 잘못된 값이 있으면', () => {
  it('에러 페이지를 띄운다', () => {
     ...
  });
});
```

jiunkoo replied:
더 명확한 표현이네요. 리팩토링 할 때 적용하도록 하겠습니다.

hannut91 commented on src/RestaurantsPage.test.jsx:
```diff
@@ -0,0 +1,38 @@
+  }));
+
+  const { queryByText } = render((
+    <MemoryRouter>
```
레스토랑 페이지에서는 MemoryRouter를 안써도 될거에요

jiunkoo replied:
MemoryRouter를 제거하니 에러가 떠서 그대로 두었습니다!

hannut91 replied:
헉 그렇군요 알겠습니다!

jiunkoo committed:
[리팩토링] 코드 리뷰를 반영하여 리팩토링
- Link to를 h1 헤더 안에 집어넣기
- App의 테스트에서 반복되는 렌더 부분을 함수로 변경
- RestaurantDetailPage의 테스트 문구를 맥락에 맞게 수정

jiunkoo committed:
[리팩토링] RestaurantDetail과 RestaurantDetailContainer 합치기

jiunkoo commented:
리팩토링이 많이 늦었네요! 코드리뷰 반영했고, RestaurantDetail 페이지가 불필요하다고 판단해서 RestaurantDetailContainer와 합쳤습니다!

hannut91 commented on src/RestaurantDetailContainer.test.jsx:
```diff
@@ -0,0 +1,67 @@
+            name: '팔보채',
+          },
+        ],
+      };
+
+      useSelector.mockImplementation((selector) => selector({
+        restaurant,
+      }));
+
+      const { getByText } = render((
+        <MemoryRouter>
+          <RestaurantDetailContainer />
+        </MemoryRouter>
+      ));
+
+      expect(getByText(/양천주가/)).not.toBeNull();
+      expect(getByText(/서울 강남구/)).not.toBeNull();
+      expect(getByText(/탕수육/)).not.toBeNull();
+      expect(getByText(/팔보채/)).not.toBeNull();
+    });
+  });
+
+  context('레스토랑 값이 화면에 없는 경우', () => {
+    it('로딩 화면을 출력한다', () => {
+      const restaurant = null;
+
+      useSelector.mockImplementation((selector) => selector({
+        restaurant,
+      }));
+
+      const { container } = render((
+        <MemoryRouter>
+          <RestaurantDetailContainer />
+        </MemoryRouter>
+      ));
+
+      expect(container).toHaveTextContent('continue loading...');
+    });
+  });
+});
```
```js
describe('RestaurantDetailContainer', () => {
  const rednerRestaurantDetailContainer = () => render(
    <MemoryRouter>
      <RestaurantDetailContainer />
    </MemoryRouter>
  );

  context('레스토랑 값이 있는 경우', () => {\
    beforeEach(() => {
      const restaurant = {
        id: 1,
        categoryId: 1,
        name: '양천주가',
        address: '서울 강남구',
        menuItems: [
          {
            id: 16,
            restaurantId: 1,
            name: '탕수육',
          },
          {
            id: 17,
            restaurantId: 1,
            name: '팔보채',
          },
        ],
      };

      useSelector.mockImplementation((selector) => selector({
        restaurant,
      }));
    });

    it('레스토랑 정보를 화면에 출력한다', () => {
      const { getByText } = rednerRestaurantDetailContainer();

      expect(getByText(/양천주가/)).not.toBeNull();
      expect(getByText(/서울 강남구/)).not.toBeNull();
      expect(getByText(/탕수육/)).not.toBeNull();
      expect(getByText(/팔보채/)).not.toBeNull();
    });
  });

  context('레스토랑 값이 화면에 없는 경우', () => {
    beforeEach(() => {
      useSelector.mockImplementation((selector) => selector({
        restaurant: null
      }));
    });

    it('로딩 화면을 출력한다', () => {
      const { container } = rednerRestaurantDetailContainer();

      expect(container).toHaveTextContent('continue loading...');
    });
  });
});
```

이렇게 하면 context에서 무엇이 달라지는지 좀 더 명확하게 표현할 수 있고 테스트 부분을 좀 더 간단하게 표현해볼 수 있을 것 같아요.

jiunkoo replied:
넵 다음 과제에서는 beforeEach를 적용해서 TDD를 작성해보겠습니당!

hannut91 commented on src/RestaurantDetailPage.test.jsx:
```diff
@@ -0,0 +1,79 @@
+          match={{
+            params: { restaurantId: restaurant.id }, isExact: true, path: '', url: '',
+          }}
+        />
+      ));
+
+      expect(dispatch).toBeCalled();
+
+      expect(queryByText(/양천주가/)).not.toBeNull();
+      expect(queryByText(/서울 강남구/)).not.toBeNull();
+      expect(queryByText(/탕수육/)).not.toBeNull();
+      expect(queryByText(/팔보채/)).not.toBeNull();
+    });
+  });
+
+  context('param에 restarauntId가 잘못된 값이 있으면', () => {
+    it('에러 페이지를 띄운다', () => {
+      const restaurant = {};
+      const dispatch = jest.fn();
+
+      useDispatch.mockImplementation(() => dispatch);
+
+      useSelector.mockImplementation((selector) => selector({
+        restaurant,
+      }));
+
+      const { queryByText } = render((
+        <RestaurantDetailPage
+          match={{
+            params: { restaurantId: 'hello' }, isExact: true, path: '', url: '',
+          }}
+        />
+      ));
+
+      expect(dispatch).not.toBeCalled();
+
+      expect(queryByText(/잘못된 접근입니다!/)).not.toBeNull();
+    });
+  });
+});
```
```js
describe('RestaurantDetailPage', () => {
  const restaurant = {
    id: 1,
    categoryId: 1,
    name: '양천주가',
    address: '서울 강남구',
    menuItems: [
      {
        id: 16,
        restaurantId: 1,
        name: '탕수육',
      },
      {
        id: 17,
        restaurantId: 1,
        name: '팔보채',
      },
    ],
  };

  const dispatch = jest.fn();

  const renderRestaurantDetailPage = (restaurantId) => render(
    <RestaurantDetailPage
      match={{
        params: { restaurantId }, isExact: true, path: '', url: '',
      }}
    />
  );

  beforeEach(() => {
    dispatch.mockClear();
    useDispatch.mockImplementation(() => dispatch);

    useSelector.mockImplementation((selector) => selector({
      restaurant,
    }));
  });

  context('param에 restaurantId가 있으면', () => {
    it('숫자인 경우 레스토랑 페이지를 띄운다', () => {
      const { queryByText } = renderRestaurantDetailPage(restaurant.id);

      expect(dispatch).toBeCalled();

      expect(queryByText(/양천주가/)).not.toBeNull();
      expect(queryByText(/서울 강남구/)).not.toBeNull();
      expect(queryByText(/탕수육/)).not.toBeNull();
      expect(queryByText(/팔보채/)).not.toBeNull();
    });
  });

  context('param에 restarauntId가 잘못된 값이 있으면', () => {
    it('에러 페이지를 띄운다', () => {
      const { queryByText } = renderRestaurantDetailPage('WRONG_ID');

      expect(dispatch).not.toBeCalled();

      expect(queryByText(/잘못된 접근입니다!/)).not.toBeNull();
    });
  });
});
```

이렇게 하면 각 테스트에서 어떤 값이 달라져서 어떻게 결과가 달라지는지 좀 더 명확하게 표현할 수 있을 것 같아요

hannut91 reviewed (approved):
고생하셨습니다! 저도 한 번 리팩터링해봐서 커멘트 남겨드렸어요 ㅎㅎ 아샬님의 과제풀이 영상을 통해 복습해보시면 좋을 것 같아요

hannut91 commented:
이번 과제를 통해 무엇을 배우셨나요? 더 배우고 싶은게 있으신가요?

jiunkoo commented:
양식에 따라서 주간회고를 하니 형식 자체에 얽매이는 것 같아서, 하나로 묶어서 남깁니다.

 6주차 과제의 질문은 '변경되는 주소 페이지를 어떻게 테스트 하는가?'였고, 해결 방법은 렌더링할 페이지를 감싸고 라우터의 주소를 바꾸는 것이었습니다. 과제를 하면서 힘들었던 건 홈에서 마우스 클릭을 통해서 링크가 넘어가는데, 주소창에 입력하면 안 넘어가는 버그가 발생했을 때입니다. webpack 설정을 했는데도 레스토랑 상세 페이지만 출력이 안 되어서, 다시 index.html 설정과 webpack 설정을 바꾸니 그제야 테스트가 잘 실행되었습니다. 거의 다 했다고 좋아했는데 마무리를 못할 때의 아찔함이란...

 매주 새로운 과제를 할 때마다 전주의 코드 리뷰를 적용해보려고 노력합니다. 이번에는 맥락에 따라 테스트가 분리되는 경우 context를 작성했고, fixture를 활용하는 대신 테스트 데이터를 직접 집어넣어 읽는 사람이 보기 편하도록 작성했습니다. 또한 특별히 피드백은 없었지만, 보기 쉽도록 커밋 문구를 바꾸었습니다. 강사님들이 더 보기 편하도록 TDD에 해당하는 커밋에는 [추가], [구현], [리팩토링] 말머리를, 그렇지 않은 커밋(package 설치, 패키지 설정, fixture에 테스트 데이터 추가 등)에는 말머리를 제외하고 작성했습니다. 강사님이 보기엔 어떠셨나요?

 TDD가 익숙하지 않을 때는 test로 작성하는 것도 어려웠습니다. 이제 describe와 test에 익숙해져서, context로 맥락을 분리하는 등 점점 상세하게 익혀가고 있습니다. 다음 주에는 beforeEach(+ given)를 활용해서 테스트 준비와 테스트를 분리해보고 싶습니다. 아직 배우는 단계인 만큼, 기본적인 걸 확실하게 익혀 제 것으로 만들고 싶습니다.

hannut91 commented:
커밋은 자신이 보기 좋은 걸로 해도 괜찮을 것 같아요. 리뷰어 입장에서는 코드로 다 파악이 가능해서요 ㅎㅎ 다음에 더 테스트를 명확하게 작성하는 방법을 훈련하시면 좋을 것 같네요!
